crates/typst/src/vm/opcodes.rs

```diff
@@ -0,0 +1,1832 @@
+
+                // Call the method.
+                let func = match &*func {
+                    Value::Func(func) => func.clone(),
+                    Value::Type(type_) => type_.constructor().at(span)?,
+                    _ => {
+                        bail!(span, "expected function, found {}", func.ty().long_name())
+                    }
+                };
+
+                // Call the function.
+                let value = func.call(engine, args)?;
+
+                // Write the value to the output.
+                vm.write_one(self.out, value).at(span)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Run for Field {
+    fn run(
+        &self,
+        _: &[u8],
+        span: Span,
+        vm: &mut VMState,
+        _: &mut Engine,
+    ) -> SourceResult<()> {
+        // Get the value.
+        let value = vm.read(self.access).at(span)?.read(span, vm)?;
+
+        // Write the value to the output.
+        vm.write_one(self.out, value.into_owned()).at(span)?;
+
+        Ok(())
+    }
+}
+
+impl Run for While {
+    fn run(
+        &self,
+        instructions: &[u8],
+        span: Span,
+        vm: &mut VMState,
+        engine: &mut Engine,
+    ) -> SourceResult<()> {
+        debug_assert!(self.len as usize <= instructions.len());
+
+        // SAFETY: The instruction pointer is always within the bounds of the
+        // instruction list.
+        // JUSTIFICATION: This avoids a bounds check on every scope.
+        let instructions = unsafe {
+            std::slice::from_raw_parts(instructions.as_ptr(), self.len as usize)
+        };
+
+        let defaults = vm.read(self.scope).at(span)?.clone();
+
+        let f = move || {
+            let flow = vm.enter_scope(
+                engine,
+                &defaults,
+                instructions,
+                Some(Box::new(std::iter::empty())),
+                None,
+                true,
+                false,
+                span,
+            )?;
+
+            let joined = match flow {
+                ControlFlow::Done(value) => value,
+                ControlFlow::Break(_) | ControlFlow::Continue(_) => {
+                    unreachable!("unexpected control flow")
+                }
+                ControlFlow::Return(value) => {
+                    vm.state |= State::RETURNING;
+                    value
+                }
+            };
+
+            eprintln!("{:4} => Exit: {:?} + {}", self.len, joined, self.len);
+
+            if let Some(out) = self.out.ok() {
+                // Write the output to the output register.
+                vm.write_one(out, joined).at(span)?;
+            }
+
+            vm.instruction_pointer += self.len as usize;
+
+            Ok(())
+        };
+
+        // Stacker is broken on WASM.
+        #[cfg(target_arch = "wasm32")]
+        return f();
+
+        #[cfg(not(target_arch = "wasm32"))]
+        stacker::maybe_grow(32 * 1024, 2 * 1024 * 1024, f)
```

Comment:
does `while` loop use a lot of stack?

Reply:
It create a scope which allocates ~10 kB of cache space, I don't know (yet) whether the stackers are needed, I'll see as I go along